plugin.whereareyou/resources/lib/whereareyou.py

```diff
@@ -1,48 +1,179 @@
+from kodi_six import xbmc, xbmcgui
 from kodi_six.utils import py2_decode
-import os, sys
+import json, os, sys
+from resources.lib.waysettings import loadSettings
+from resources.lib.xlogger import Logger
+from resources.lib.api.harmony import HubControl
 try:
     from urllib.parse import unquote_plus as _unquote_plus
 except ImportError:
     from urllib import unquote_plus as _unquote_plus
 
-addon        = xbmcaddon.Addon()
-addonname    = addon.getAddonInfo('id')
-addonversion = addon.getAddonInfo('version')
-addonpath    = addon.getAddonInfo('path')
-addonicon    = xbmc.translatePath('%s/icon.png' % addonpath )
-language     = addon.getLocalizedString
-preamble     = '[Where Are You]'
+def _upgrade():
+    settings = loadSettings()
+    if settings['version_upgrade'] != settings['ADDONVERSION']:
+        settings['ADDON'].setSetting( 'version_upgrade', settings['ADDONVERSION'] )
 
 
 
 class Main:
 
     def __init__( self ):
+        self._init_vars()
+        self.LW = Logger( preamble='[Where Are You]', logdebug=self.SETTINGS['debug'] )
+        self.LW.log( ['script version %s started' % self.SETTINGS['ADDONVERSION']], xbmc.LOGINFO )
         self._parse_argv()
         if self.TITLE and self.MESSAGE:
-            dialog = xbmcgui.Dialog()
-            ok = dialog.ok( self.TITLE, self.MESSAGE )
-            self.play_video( os.path.join( addonpath, 'resources', 'blank.mp4' ) )
+            self._display_dialog()
+        elif self.SETTINGS['harmonycontrol']:
+            self._mappings_options()
+            if self.FROMSETTINGS:
+                self.SETTINGS['ADDON'].openSettings()    
+        self.LW.log( ['script stopped'], xbmc.LOGINFO )
+
+
+    def _display_dialog( self ):
+        use_extended_dialog = False
+        try:
+            json_mappings = json.loads( self.SETTINGS['mappings'] )
+        except ValueError:
+            json_mappings = {}
+        if json_mappings and self.SETTINGS['harmonycontrol']:
+            for item in json_mappings:
+                thematch = json_mappings.get( item, {} ).get( 'match', '' )
+                self.LW.log( ['checking for %s in %s' % (thematch, self.TITLE)] )
+                self.LW.log( ['checking for %s in %s' % (thematch, self.MESSAGE)] )
+                if thematch.lower() in self.TITLE.lower() or thematch.lower() in self.MESSAGE.lower():
+                    self.LW.log( ['found match'] )
+                    activity, cmds = self._get_mapping_details( json_mappings, item )
+                    self.MESSAGE = '%s\n%s' % (self.MESSAGE, self.SETTINGS['ADDONLANGUAGE']( 32303 ))
+                    use_extended_dialog = True
+                    break
+        if use_extended_dialog:
+            if self.DIALOG.yesno( self.TITLE, self.MESSAGE ):
+                self._run_activity( activity, cmds )
+        else:
+            self.DIALOG.ok( self.TITLE, self.MESSAGE )
+        xbmc.Player().play( os.path.join( self.SETTINGS['ADDONPATH'], 'resources', 'blank.mp4' ) )
```

Comment:
Would be better to have this point to a plugin:// path that could later be resolved

Reply:
Unfortunately if I change that line to use plugin://<scriptid>/resources/blank.mp4 it crashes Kodi (although there is nothing obvious in the logs as to why). I'm open to suggestions on how to address that.  You might have noticed in the original version I used xbmcplugin.setResolvedUrl() to set a path to play, and that works if you're in library mode. But if you do it from the home screen, Kodi ends up in a loop constantly calling the addon over and over. It looks like the home screen call generates a playlist (the library mode doesn't), and once an item is in that playlist, it just keeps playing it.  I tried like 10 different things to deal with this, and the only one that mostly works is what I submitted. This still generates a "next item in playlist not found" message on the home screen, but it's better than the alternatives. I'm only playing the video at all so Kodi doesn't complain that it couldn't play the video.  I guess this is what happens when you try and find a work around for stub files not showing messages with no optical drive (which I know is fixed in Matrix, but this addon now has some extra options, so I think it's still potentially useful in Matrix). '-)

Reply:
thats fine for now. Kodi definitely needs a lot of work on the playlist handling